src/file.cpp

```diff
@@ -1426,6 +1426,13 @@ namespace libtorrent
 			| FILE_FLAG_OVERLAPPED
 			| ((mode & no_cache) ? FILE_FLAG_WRITE_THROUGH : 0);
 
+		if ((mode & sparse) == 0)
+		{
+			// Enable privilege required by SetFileValidData()
+			// https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-setfilevaliddata
+			get_manage_volume_privs();
```

Comment:
Is there a good reason to move the "call once" logic into the `get_manage_volume_privs()` function?
I would prefer the keep this patch as small as possible. Something like just initializing a `static` variable here with its call. I'm also not sure it's a good idea to condition this on whether the `sparse` flag is set or not. It could be unset later, which would complicate the logic for calling this function.

I think it would be better (and simpler) to keep `get_manage_volume_privs()` as it is, and just make sure it's called once in the constructor.

Reply:
OK, I'll remove that condition.

Reply:
> Is there a good reason to move the "call once" logic into the get_manage_volume_privs() function?

Since I'm not sure that this function should really be called only once, moving that logic to the function implementation makes sense. Docs tell us that `SetValidFileData()` require file handle to be initially opened with `SeManageVolumePrivilege` acquired. If we enable it once, there is no guaranty that other code won't disable this privilege afterwards. Thus `get_manage_volume_privs()` is called each time before `CreateFile`, but the hidden 'call once' logic makes it work as before.

Reply:
Also I believe that `(mode & sparse) == 0` condition is correct and should persist, because file just can't change sparse mode on the fly.

Reply:
that's a good point. You don't think calling it on every file-open will add an unacceptable slow down?

Reply:
Everything is possible. For now it won't because of hidden call-once logic. Disabling that logic is definitely requires further consideration, that is out of current scope.

Also out of scope, but I think that using that API should be restricted by introducing special flag, that the user must explicitly specify. In such case user becomes responsible for related security flaws.

Reply:
the security risks are mitigated somewhat by libtorrent itself tracking which parts of a file has been written, and won't serve up "garbage" data (unless coerced by invalid resume data). It turns out people run bittorrent clients as administrator just to get access to this feature.

Reply:
Did you consider the case when application crash or spontaneous reboot happens after that API call but before libtorrent manage to write any data to file. Any user will be able to explore sensible data that previously belonged to some file that was removed.

Reply:
yes, in all such cases the resume file will indicate that we don't have those pieces, and they won't be advertised to peers and requests for them will be rejected.

Reply:
No, that's not what I worry about. What you call a "garbage data" potentially could be a wide variety of sensitive information such as stored passwords, crypto-keys, etc. All this could become visible to any local user with limited permissions. Someone would call it a __vulnerability__.

Reply:
right, if the person with administrator privileges make the file accessible to other users. Currently you need to run the client with administrator access though, which hasn't been the default since XP, so it's quite deliberate.

Reply:
I agree. I think it's bonkers that Microsoft still support this feature